src/modules/rust.rs

```diff
@@ -1,601 +1,923 @@
+        formatter
+            .map_meta(|var, _| match var {
+                "symbol" => Some(config.symbol),
+                _ => None,
+            })
+            .map_style(|variable| match variable {
+                "style" => Some(Ok(config.style)),
+                _ => None,
+            })
+            .map(|variable| match variable {
+                "version" => get_module_version(context, &config, &rust_env_info).map(Ok),
+                "numver" => get_module_numeric_version(context, &config, &rust_env_info).map(Ok),
+                "toolchain" => get_toolchain_version(context, &config, &rust_env_info).map(Ok),
+                _ => None,
+            })
+            .parse(None, Some(context))
+    });
+
+    module.set_segments(match parsed {
+        Ok(segments) => segments,
+        Err(error) => {
+            log::warn!("Error in module `rust`:\n{}", error);
+            return None;
+        }
+    });
+
+    Some(module)
+}
+
+fn get_module_version(
+    context: &Context,
+    config: &RustConfig,
+    rust_env_info: &RustToolingEnvironmentInfo,
+) -> Option<String> {
+    type Outcome = RustupRunRustcVersionOutcome;
+
+    match rust_env_info.get_rustup_rustc_version(context) {
+        Outcome::RustcVersion(rustc_version) => {
+            format_rustc_version(rustc_version, config.version_format)
+        }
+        Outcome::RustupNotWorking | Outcome::ToolchainUnknown => {
+            // If `rustup` can't be executed, or there is no environmental toolchain, we can
+            // execute `rustc --version` without triggering a toolchain download
+            format_rustc_version(&execute_rustc_version(context)?, config.version_format)
+        }
+        Outcome::ToolchainNotInstalled(name) => Some(name.to_string()),
+        Outcome::Err => None,
+    }
+}
+
+fn get_module_numeric_version(
+    context: &Context,
+    _config: &RustConfig,
+    rust_env_info: &RustToolingEnvironmentInfo,
+) -> Option<String> {
+    type Outcome = RustupRunRustcVersionOutcome;
+
+    match rust_env_info.get_rustup_rustc_version(context) {
+        Outcome::RustcVersion(version) => {
+            let release = version.split_whitespace().nth(1).unwrap_or(version);
+            Some(format_semver(release))
+        }
+        Outcome::RustupNotWorking | Outcome::ToolchainUnknown => {
+            let (numver, _toolchain) = rust_env_info.get_rustc_verbose_version(context)?;
+            Some(numver.to_string())
+        }
+        Outcome::ToolchainNotInstalled(_) | RustupRunRustcVersionOutcome::Err => None,
+    }
+}
+
+fn get_toolchain_version(
+    context: &Context,
+    _config: &RustConfig,
+    rust_env_info: &RustToolingEnvironmentInfo,
+) -> Option<String> {
+    type Outcome = RustupRunRustcVersionOutcome;
+
+    let default_host_triple = rust_env_info
+        .get_rustup_settings(context)
+        .default_host_triple()
+        .or_else(|| guess_host_triple());

```

Comment:
```suggestion
        .or_else(guess_host_triple);
```

Reply:
Somehow, there's a lifetime error that happens when you do that. I don't have the motivation to debug lifetime issues on chained methods at the moment, which is why I have this [weird set of assignments](https://github.com/starship/starship/pull/3414/files#diff-9916a3a9c5b9f49331e4569e3ef2d228d451eb39a98d44fcc37df570534b730fR233) in the current commit, but that's definitely something to tackle later.